### ISSUE TITLE
4.0.0 serialization of external types

### ISSUE THREAD

**Issue by RbsTechOnline** [dotnet/orleans#7635]
I have been doing a lot of work with 4.0.0, its definitely a new world and impressive.
 
However, I am sharing types between my API -> Orleans layer and Blazor through a shared DTO class library. 

How do i make this type known to the Orleans Serilaizer? 

I am trying to get keep as much as i can out of the Shared Libs for obvious wasm reasons. What are the external serialization options with 4.0.0?

**Comment by RbsTechOnline** [comment]
I tried every attribute i could, searched every test, and trawled through the generator, i stopped short of debugging the code generator. However to solved the solution and just set up low tech build macro to copy classes for now, its enough for me to keep trialing the preview code and writing the plugins i need. Let me know if i am missing something obvious with support for library types thanks

**Comment by ReubenBond** [comment]
What was the codegen issue, @RbsTechOnline? We need to improve this use case. The idea is that we will allow you to define surrogate types. I assume you're already aware of `[GenerateSerializer]` and that it's not suitable for your case.

You can generate serializers for types in an external DLL by using this attribute in a project which has the code generator installed (implying that you do not install it in the project your types are defined in):
``` C#
[assembly: GenerateCodeForDeclaringAssembly(typeof(MyPocoType)]
```

**Comment by RbsTechOnline** [comment]
Hi @ReubenBond  I had the `[GenerateSerializer]` attribute on the DTO in the external library. The Compiler was just complaining that it couldn't find the namespace of the dto in the `g.cs` file `CS0234`. 

However, circling back around to the `GenerateCodeForDeclaringAssembly` as you suggested, it turns out it was the attribute i was looking for and working. i got bit distracted by known types and application parts, and must have munted this implementation when i checked it the first time. Obviously the documentation will cover this in the future, and possibly some warning for future travelers, through i have nothing concrete to suggest at this stage. 

Just for future reference, i found a similar issue after the fact

https://github.com/dotnet/orleans/issues/7593

Anyway thanks for your help.

**Comment by ReubenBond** [comment]
> The Compiler was just complaining that it couldn't find the namespace of the dto in the g.cs file CS0234.

I think this was likely caused by a recently fixed bug: #7753

**Comment by ReubenBond** [comment]
Opened a PR to support for serialization surrogates: https://github.com/dotnet/orleans/pull/7776
